### PR TITLE
fix the constructor expr in the exported structure's struct-info

### PR DESCRIPTION
**Pull request by capfredf**

### CODE REVIEW

**Comment by samth** [comment]
Can you write a test for this?

File: typed-racket-lib/typed-racket/typecheck/def-binding.rkt [hunk 1]
```diff
@@ -116,17 +116,29 @@
...
 
      (define/with-syntax (constr* type-desc* pred* super* accs* ...)
        (for/list ([i (in-list (cons constr-new-id new-ids))])
          (and (identifier? i) #`(quote-syntax #,i))))
 
+     (define/with-syntax constr-in-si*
+       (if (free-identifier=? new-id constr-new-id)
+           new-id
+           constr-new-id))
+
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/def-binding.rkt:131]: Why are we defining this twice?

Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/def-binding.rkt]: good catch

File: typed-racket-test/succeed/test-struct-info.rkt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  (check-constr kiwi kiwi)
+  (check-constr apple make-apple)
+  (check-constr pear pear)
+  (check-constr dragon-fruit make-dragon-fruit)
+  (check-constr DragonFruit make-dragon-fruit)
+  (check-constr PEAR pear))
+
+(require 'typed)
+(require 'test)
+(check-constr apple make-apple)
```
Comment (samth) [inline typed-racket-test/succeed/test-struct-info.rkt:35]: I feel like this test could be simpler/more direct. For example, is `(object-name apple)` always `'make-apple` here? Or can we just check that everything works correctly as a procedure?

Comment (capfredf) [reply typed-racket-test/succeed/test-struct-info.rkt:35]: Do you mean we use `object-name` at phase 0? my understanding is 
1. if v is simply a structure type identifier and not a constructor name, then we will see racket/typed racket complains identifier for static struct-type information cannot be used as an expression before `(object-name v)` run
2. `object-name` doesn't return the constructor name. E.g `(object-name (make-apple))` will return `'apple`
3. If I understand you correctly, everything already almost worked correctly as a procedure before this patch. Previously we used `export-id` as the constructor name. In this case, the struct info of `(struct apple () #:constructor-name make-apple)`  incorrectly had some form of #'apple as the constructor. But we found no errors from running TR's tests, buiding packages and running their tests until `delay-pure` which directly used the struct info of typed structure types.

Comment (samth) [reply typed-racket-test/succeed/test-struct-info.rkt:35]: I thinking of this code:

```
#lang racket

(module test racket/base
  (require  (for-syntax racket/base syntax/parse racket/struct-info racket/match))

  (define-syntax (check-constr stx)
    (syntax-parse stx
      [(_ sname:id)
       (define si (extract-struct-info (syntax-local-value #'sname (lambda () #f))))
       (match-define (list* struct-desc constr rst) si)
       constr]))
  (provide check-constr))
(module t typed/racket/base
  (require (submod ".." test) typed/rackunit)
  (struct kiwi ())
  (struct apple () #:constructor-name make-apple)
  (struct pear () #:type-name PEAR)
  (struct dragon-fruit () #:type-name DragonFruit #:constructor-name make-dragon-fruit)

  (check-equal? (check-constr apple) 'make-apple)

  (provide (struct-out apple)
           (struct-out kiwi)
           (struct-out PEAR)
           (struct-out DragonFruit)))

(require 'test 't rackunit)
(check-equal? (check-constr apple) 'make-apple)
```

Comment (capfredf) [reply typed-racket-test/succeed/test-struct-info.rkt:35]: Ah, this is better

File: typed-racket-test/succeed/test-struct-info.rkt [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#lang racket/base
+(module test racket/base
+  (require (for-syntax racket/base syntax/parse racket/struct-info racket/match))
+
+  (define-syntax (get-constr stx)
+    (syntax-parse stx
+      [(_ sname:id)
+       (define si (extract-struct-info (syntax-local-value #'sname (lambda () #f))))
+       (match-define (list* struct-desc constr rst) si)
+       #`'#,(syntax-e constr)]))
```
Comment (samth) [inline typed-racket-test/succeed/test-struct-info.rkt:10]: Why `syntax-e`? Can't you just use `constr` here?

Comment (samth) [reply typed-racket-test/succeed/test-struct-info.rkt:10]: I guess you are trying to see the _name_ of the binding that is specified, but I think we shouldn't care about that, and should care about it having the right name.

Comment (capfredf) [reply typed-racket-test/succeed/test-struct-info.rkt:10]: then for `check-equal?`, we would be using its runtime value, which is a procedure

**Review by samth — APPROVED** [review]
Good when tests pass.

**Pull request merged**